### PR TITLE
[python][torch] Support distributed iterable dataset sharding

### PR DESCRIPTION
**Pull request by XiaoHongbo-Hope**
## What changed

- Shard streaming Torch datasets across DDP ranks and DataLoader workers when `auto_detect_rank=True`.
- Keep rank sharding disabled by default because `to_torch` accepts caller-planned splits.
- Resolve rank from `torch.distributed`, then torchrun environment variables, and preserve it in spawned workers.
- Reject binding limits with multiple ranks.
- Keep shuffled reads reproducible per epoch, rank, and worker.
- Document uneven inputs and `DistributedDataParallel.join()`.

## Reference

The API is inspired by [LanceDataset](https://github.com/lance-format/lance/blob/main/python/python/lance/torch/data.py). Unlike Lance, this low-level API accepts external splits, so automatic sharding is opt-in.

## Validation

- 43 Torch tests passed.
- CPU torchrun test executes DDP forward/backward with 2 ranks and 2 spawned DataLoader workers.
- Split assignment is complete and non-overlapping.
- Pre-sharded splits remain unchanged by default.
- flake8 and `git diff --check` passed.

### CODE REVIEW

**Review by JingsongLi — COMMENTED** [review]
Found two issues in the distributed Torch sharding path; please see the inline comments.

File: docs/docs/pypaimon/pytorch.md [hunk 1]
```diff
@@ -60,6 +60,30 @@ when it is false, it will read the full amount of data into memory.
...
+
+Streaming reads shard splits across DDP ranks and DataLoader workers:
+
+```python
+dataset = table_read.to_torch(
+    splits,
+    streaming=True,
+    auto_detect_rank=True,
+)
+dataloader = DataLoader(dataset, batch_size=32, num_workers=2)
```
Comment (JingsongLi) [inline docs/docs/pypaimon/pytorch.md]: **[P1] Use a spawn-safe DataLoader in this DDP example**

On Linux, leaving `multiprocessing_context` unset makes a multi-worker DataLoader use `fork`. PyTorch explicitly warns that DDP with NCCL (and Gloo over InfiniBand) is not fork-safe and is likely to deadlock. This is especially easy to miss because the new torchrun test uses `multiprocessing_context="spawn"`, so it does not exercise the documented path. Please set `multiprocessing_context="spawn"` (or `forkserver`) in the example and mention the usual guarded-entry-point requirement for spawn. See https://docs.pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html.

File: paimon-python/pypaimon/read/datasource/torch_dataset.py [hunk 2]
```diff
@@ -40,6 +41,59 @@ def _share_epoch_with_torch_workers(value):
...
+    if not auto_detect_rank:
+        return 0, 1
+
+    distributed = getattr(torch, "distributed", None)
+    if (
+        distributed is not None
+        and distributed.is_available()
+        and distributed.is_initialized()
+    ):
+        rank = distributed.get_rank()
```
Comment (JingsongLi) [inline paimon-python/pypaimon/read/datasource/torch_dataset.py:85]: **[P2] Shard against the DDP process group, not always the global world**

Calling `get_rank()` and `get_world_size()` without a group always resolves the default global process group, while DDP may be constructed with `process_group=subgroup`. In subgroup or hybrid-parallel training, `auto_detect_rank=True` will therefore shard inputs across the global world instead of the replicas synchronized by DDP; a subgroup can see only part of its dataset, or model-parallel ranks can receive different inputs. Please either accept explicit sharding rank/world size resolved from the intended DDP group before serializing the dataset to workers, or document that auto-detection supports only whole-world DDP and require caller-side pre-sharding otherwise. See https://docs.pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html.

**Comment by JingsongLi** [comment]
I compared this design with the current PyTorch, Lance, Hugging Face Datasets, WebDataset, and Ray Data implementations. Overall, the direction in this PR looks sound, and I do not see an architectural blocker.

The main reasons are:

* A plain PyTorch `IterableDataset` must own worker-replica sharding itself (`get_worker_info()`); DDP does not shard inputs, and `DataLoader` rejects a custom sampler for an `IterableDataset`. Therefore, doing process-rank sharding in the dataset and then DataLoader-worker sharding is the correct layer. See the [PyTorch DataLoader source](https://github.com/pytorch/pytorch/blob/main/torch/utils/data/dataloader.py#L359-L369) and [DDP documentation](https://docs.pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html).
* Hugging Face Datasets and WebDataset use the same broad hierarchy: first select the rank/node-local sources, then split those sources among DataLoader workers before opening/decoding them. See [HF distributed.py](https://github.com/huggingface/datasets/blob/main/src/datasets/distributed.py), [HF iterable_dataset.py](https://github.com/huggingface/datasets/blob/main/src/datasets/iterable_dataset.py), and [WebDataset shardlists.py](https://github.com/webdataset/webdataset/blob/main/src/webdataset/shardlists.py).
* Keeping `auto_detect_rank=False` by default is appropriate for Paimon. `TableRead.to_torch` receives an already-planned split list and cannot know whether the caller has pre-sharded it. Automatically sharding by default could therefore silently apply sharding twice. Systems such as Ray can safely auto-split because the trainer owns the complete dataset plan and all consumers; that is a different ownership boundary. See [Ray's dataset-shard API](https://docs.ray.io/en/latest/train/api/doc/ray.train.get_dataset_shard.html).
* The explicit `sharding_rank` / `sharding_world_size` pair is important for DDP subgroups and hybrid parallelism. Resolving these values in the parent and serializing scalar context into spawned workers is also a good design.

One useful comparison is Lance. Lance flattens distributed rank and DataLoader worker id into a global rank and supports fragment- and batch-level sharding. However, based on the current source, its distributed helper only observes a process group initialized in the current process. A spawned DataLoader worker does not retain the parent's initialized process group, while the deprecated explicit `rank/world_size` branch does not additionally shard among worker replicas. Therefore, as a source-level inference, Lance's iterable path does not transparently cover the full `spawn + DDP + num_workers > 0` combination. Paimon's captured parent rank followed by a separate worker split handles this combination more robustly. See [LanceDataset](https://github.com/lance-format/lance/blob/main/python/python/lance/torch/data.py), [Lance distributed helpers](https://github.com/lance-format/lance/blob/main/python/python/lance/torch/dist.py), and [Lance samplers](https://github.com/lance-format/lance/blob/main/python/python/lance/sampler.py).

I would keep the current exact-coverage policy as well: assign every split once, allow uneven rank lengths, and require DDP uneven-input handling. This is preferable to silently duplicating or dropping data. A few non-blocking follow-ups may still be worthwhile:

1. Clarify that `set_epoch()` changes each rank's local buffer-shuffle order, but does not reshuffle split ownership across ranks as PyTorch `DistributedSampler` does.
2. Mention that training loops with additional non-DDP collectives (for example, `SyncBatchNorm`) should use `model.join(throw_on_early_termination=True)`; plain `join()` can only shadow DDP's own collectives.
3. Consider warning when the number of splits is smaller than `world_size * num_workers`, since some ranks/workers will be idle.
4. If equal-step training becomes a requirement later, add an explicit policy such as preserve-all/drop-to-equal/repeat-to-equal rather than changing the current behavior implicitly. Similarly, fragment/split versus batch/weighted sharding can be added as an explicit strategy later without coupling it to rank detection.

So my recommendation is to keep the current API boundary and rank -> worker -> reader hierarchy. The comparison with other implementations reinforces this design rather than suggesting that Paimon should copy Lance's defaults.

**Comment by XiaoHongbo-Hope** [comment]
> So my recommendation is to keep the current API boundary and rank -> worker -> reader hierarchy. The comparison with other implementations reinforces this design rather than suggesting that Paimon should copy Lance's defaults.

Agreed, I also went back and forth on Lance’s sharding model before settling on this design.

**Review by JingsongLi — APPROVED** [review]
Reviewed the final head 05cbebc6. The previously identified spawn-safety and DDP subgroup sharding issues are fixed, all required checks pass, and the cross-implementation comparison found no remaining architectural blocker.
